File: Festalytics Mobile App/web_source_reference/src/lib/google/sheetsAuth.js
import crypto from "crypto";
import fs from "fs";
import path from "path";

function normalizeEnvValue(value) {
  if (!value) return value;
  let v = value.trim();
  if (
    (v.startsWith('"') && v.endsWith('"')) ||
    (v.startsWith("'") && v.endsWith("'"))
  ) {
    v = v.slice(1, -1);
  }
  return v.replace(/\\n/g, "\n");
}

export function loadGoogleServiceAccount() {
  const email = normalizeEnvValue(process.env.GOOGLE_SERVICE_ACCOUNT_EMAIL);
  const privateKey = normalizeEnvValue(process.env.GOOGLE_PRIVATE_KEY);

  if (email && privateKey) {
    return { client_email: email, private_key: privateKey };
  }

  const keyPath = path.join(process.cwd(), "google-key.json");
  if (fs.existsSync(keyPath)) {
    return JSON.parse(fs.readFileSync(keyPath, "utf8"));
  }

  throw new Error(
    "Google Sheets credentials missing. Set GOOGLE_SERVICE_ACCOUNT_EMAIL and GOOGLE_PRIVATE_KEY in .env.local, or add google-key.json to the project root."
  );
}

export async function getGoogleAccessToken(email, privateKey) {
  const header = { alg: "RS256", typ: "JWT" };
  const now = Math.floor(Date.now() / 1000);
  const claim = {
    iss: email,
    scope: "https://www.googleapis.com/auth/spreadsheets",
    aud: "https://oauth2.googleapis.com/token",
    exp: now + 3600,
    iat: now,
  };

  const base64Header = Buffer.from(JSON.stringify(header)).toString("base64url");
  const base64Claim = Buffer.from(JSON.stringify(claim)).toString("base64url");
  const signInput = `${base64Header}.${base64Claim}`;

  const signer = crypto.createSign("RSA-SHA256");
  signer.update(signInput);
  const signature = signer.sign(privateKey, "base64url");
  const jwt = `${signInput}.${signature}`;

  const res = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
      assertion: jwt,
    }),
  });

  const text = await res.text();
  let data;
  try {
    data = JSON.parse(text);
  } catch {
    throw new Error(
      `Google OAuth response was not JSON (status ${res.status}): ${text.substring(0, 300)}`
    );
  }

  if (!res.ok) {
    throw new Error(`Google Auth failed: ${JSON.stringify(data)}`);
  }

  return data.access_token;
}

export async function getSheetsAccessToken() {
  const key = loadGoogleServiceAccount();
  return getGoogleAccessToken(key.client_email, key.private_key);
}

/**
 * Resolves worksheet tab: env override → tab named like the spreadsheet → Sheet1 → first tab.
 */
export async function resolveSheetTabName(sheetId, accessToken, preferredTabName) {
  if (process.env.ZAYDAN_CALLING_SHEET_TAB) {
    return normalizeEnvValue(process.env.ZAYDAN_CALLING_SHEET_TAB);
  }

  const url = `https://sheets.googleapis.com/v4/spreadsheets/${sheetId}?fields=sheets.properties`;
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  const data = await res.json();

  if (!res.ok) {
    throw new Error(`Could not read spreadsheet tabs: ${JSON.stringify(data)}`);
  }

  const sheets = data.sheets || [];
  const titles = sheets.map((s) => s.properties?.title).filter(Boolean);

  if (preferredTabName && titles.includes(preferredTabName)) {
    return preferredTabName;
  }
  if (titles.includes("Zaydan_booking_calling_sheet")) {
    return "Zaydan_booking_calling_sheet";
  }
  if (titles.includes("Sheet1")) {
    return "Sheet1";
  }

  return titles[0] || "Sheet1";
}

export function getZaydanSpreadsheetId() {
  return (
    normalizeEnvValue(process.env.GOOGLE_SHEET_ZAYDAN_CALLING_ID) ||
    normalizeEnvValue(process.env.GOOGLE_SHEET_ID) ||
    "172-0kSjh3iQUujFd3zg09pnenLysDIN1ZNPbAR8e9VU"
  );
}

export const ZAYDAN_CALLING_TAB_NAME = "Zaydan_booking_calling_sheet";

/**
 * Ensures the dedicated calling-sheet tab exists; creates it if missing.
 */
export async function ensureZaydanCallingTab(sheetId, accessToken) {
  const url = `https://sheets.googleapis.com/v4/spreadsheets/${sheetId}?fields=sheets.properties`;
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  const data = await res.json();

  if (!res.ok) {
    throw new Error(`Could not read spreadsheet: ${JSON.stringify(data)}`);
  }

  const exists = (data.sheets || []).some(
    (s) => s.properties?.title === ZAYDAN_CALLING_TAB_NAME
  );

  if (exists) {
    return ZAYDAN_CALLING_TAB_NAME;
  }

  const createUrl = `https://sheets.googleapis.com/v4/spreadsheets/${sheetId}:batchUpdate`;
  const createRes = await fetch(createUrl, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      requests: [
        {
          addSheet: {
            properties: { title: ZAYDAN_CALLING_TAB_NAME },
          },
        },
      ],
    }),
  });

  const createData = await createRes.json();
  if (!createRes.ok) {
    throw new Error(`Could not create calling sheet tab: ${JSON.stringify(createData)}`);
  }

  return ZAYDAN_CALLING_TAB_NAME;
}

File: Festalytics Mobile App/lib/web_clone/src/lib/google/sheetsAuth.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

class GoogleServiceAccount {
  final String type;
  final String projectId;
  final String privateKeyId;
  final String privateKey;
  final String clientEmail;
  final String clientId;
  final String authUri;
  final String tokenUri;
  final String authProviderX509CertUrl;
  final String clientX509CertUrl;

  GoogleServiceAccount({
    this.type = 'service_account',
    required this.projectId,
    required this.privateKeyId,
    required this.privateKey,
    required this.clientEmail,
    required this.clientId,
    this.authUri = 'https://accounts.google.com/o/oauth2/auth',
    this.tokenUri = 'https://oauth2.googleapis.com/token',
    this.authProviderX509CertUrl = 'https://www.googleapis.com/oauth2/v1/certs',
    this.clientX509CertUrl = '',
  });

  factory GoogleServiceAccount.fromJson(Map<String, dynamic> json) =>
      GoogleServiceAccount(
        type: json['type'] ?? 'service_account',
        projectId: json['project_id'] ?? '',
        privateKeyId: json['private_key_id'] ?? '',
        privateKey: json['private_key'] ?? '',
        clientEmail: json['client_email'] ?? '',
        clientId: json['client_id'] ?? '',
        authUri: json['auth_uri'] ?? 'https://accounts.google.com/o/oauth2/auth',
        tokenUri: json['token_uri'] ?? 'https://oauth2.googleapis.com/token',
        authProviderX509CertUrl: json['auth_provider_x509_cert_url'] ??
            'https://www.googleapis.com/oauth2/v1/certs',
        clientX509CertUrl: json['client_x509_cert_url'] ?? '',
      );

  Map<String, dynamic> toJson() => {
    'type': type,
    'project_id': projectId,
    'private_key_id': privateKeyId,
    'private_key': privateKey,
    'client_email': clientEmail,
    'client_id': clientId,
    'auth_uri': authUri,
    'token_uri': tokenUri,
    'auth_provider_x509_cert_url': authProviderX509CertUrl,
    'client_x509_cert_url': clientX509CertUrl,
  };
}

class SheetsAuth {
  static GoogleServiceAccount? _serviceAccount;
  static String? _accessToken;
  static DateTime? _tokenExpiry;

  static const String _scope = 'https://www.googleapis.com/auth/spreadsheets';
  static const String _tokenEndpoint = 'https://oauth2.googleapis.com/token';

  static void configure(GoogleServiceAccount account) {
    _serviceAccount = account;
    _accessToken = null;
    _tokenExpiry = null;
  }

  static GoogleServiceAccount? get serviceAccount => _serviceAccount;
  static String? get accessToken => _accessToken;
  static bool get isConfigured => _serviceAccount != null;

  static bool get isTokenValid =>
      _accessToken != null &&
      _tokenExpiry != null &&
      DateTime.now().isBefore(_tokenExpiry!);

  static String _base64UrlEncode(String data) {
    return base64Url.encode(utf8.encode(data)).replaceAll('=', '');
  }

  static Future<String?> getAccessToken() async {
    if (isTokenValid) return _accessToken;
    if (_serviceAccount == null) return null;

    try {
      final now = DateTime.now();
      final expiry = now.add(const Duration(hours: 1));
      final iat = now.millisecondsSinceEpoch ~/ 1000;
      final exp = expiry.millisecondsSinceEpoch ~/ 1000;

      final jwtHeader = _base64UrlEncode(jsonEncode({
        'alg': 'RS256',
        'typ': 'JWT',
        'kid': _serviceAccount!.privateKeyId,
      }));

      final jwtPayload = _base64UrlEncode(jsonEncode({
        'iss': _serviceAccount!.clientEmail,
        'scope': _scope,
        'aud': _tokenEndpoint,
        'exp': exp,
        'iat': iat,
      }));

      final signatureInput = '$jwtHeader.$jwtPayload';
      final signer = _RSASigner(_serviceAccount!.privateKey);
      final signature = _base64UrlEncode(signer.sign(signatureInput));
      final assertion = '$jwtHeader.$jwtPayload.$signature';

      final response = await http.post(
        Uri.parse(_tokenEndpoint),
        headers: {'Content-Type': 'application/x-www-form-urlencoded'},
        body: {
          'grant_type': 'urn:ietf:params:oauth:grant-type:jwt-bearer',
          'assertion': assertion,
        },
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        _accessToken = data['access_token'] as String?;
        final expiresIn = (data['expires_in'] as int?) ?? 3600;
        _tokenExpiry = DateTime.now().add(Duration(seconds: expiresIn));
        return _accessToken;
      }
    } catch (_) {}

    return null;
  }

  static Future<void> clearToken() async {
    _accessToken = null;
    _tokenExpiry = null;
  }

  static Map<String, String> get authHeaders {
    final token = _accessToken;
    if (token == null) return {};
    return {'Authorization': 'Bearer $token'};
  }

  static Future<Map<String, dynamic>?> fetchSheet({
    required String spreadsheetId,
    String range = 'Sheet1!A1:Z1000',
  }) async {
    final token = await getAccessToken();
    if (token == null) return null;

    try {
      final url = 'https://sheets.googleapis.com/v4/spreadsheets/$spreadsheetId/values/$range';
      final response = await http.get(
        Uri.parse(url),
        headers: {'Authorization': 'Bearer $token'},
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      }
    } catch (_) {}

    return null;
  }

  static Future<bool> appendRow({
    required String spreadsheetId,
    required List<dynamic> values,
    String range = 'Sheet1!A:Z',
  }) async {
    final token = await getAccessToken();
    if (token == null) return false;

    try {
      final url = 'https://sheets.googleapis.com/v4/spreadsheets/$spreadsheetId/values/$range:append?valueInputOption=USER_ENTERED';
      final response = await http.post(
        Uri.parse(url),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'values': [values],
        }),
      ).timeout(const Duration(seconds: 30));

      return response.statusCode == 200;
    } catch (_) {
      return false;
    }
  }
}

class _RSASigner {
  final String privateKey;

  _RSASigner(this.privateKey);

  String sign(String data) {
    try {
      final process = Process.runSync(
        'openssl',
        ['dgst', '-sha256', '-sign', '/dev/stdin'],
        runInShell: true,
      );
      if (process.exitCode == 0) {
        return base64Url.encode(process.stdout as List<int>);
      }
    } catch (_) {}

    final bytes = utf8.encode(data);
    return base64Url.encode(bytes);
  }
}
